### ISSUE TITLE
Windows support with msvc 2015 rtm

### ISSUE THREAD

**Issue by sbant** [sandstorm-io/capnproto#227]
Today, vs2015 is on the official site.

Is it enough to build capnp? (except rpc support)

**Comment by kentonv** [comment]
Sorry, I'm not sure I understand the question. Cap'n Proto has supported VS (in lite mode) since capnp 0.5.0 + vs2015 preview last December.

Is the question whether the latest release of VS2015 still works? I would guess that it does, though honestly I haven't tested it.

Is the question whether non-lite mode can be supported? The answer to that is almost certainly "no", since there is expected to be porting work even once MSVC is ready.

**Comment by sbant** [comment]
I mean if capnp can build with msvc 2015 rtm for dynamic support(pycapnp need Dynamic API, the current lite mode for windows missing Dynamic API and RPC if I'm not wrong), since kj-async is hard to port to Windows.

In this page, it says constexpr is complete in msvc 2015 rtm
http://blogs.msdn.com/b/vcblog/archive/2015/06/19/c-11-14-17-features-in-vs-2015-rtm.aspx


But I just try build capnp with msvc 2015 rtm(set CAPNP_LITE 0),
found some constexpr still can't build in "kj", and also many errors in "capnp".

**Comment by kentonv** [comment]
Yeah unfortunately it's going to take some porting effort to support the dynamic API in MSVC even if MSVC has perfect C++11 support (which I believe it still doesn't). I can plan on working on this whenever we get around to a 0.6 release but I can't make any promises on when that might be.

**Comment by scottpurdy** [comment]
+1 For Windows support for Cap'n Proto and pycapnp. I may be able to help if there are some concrete tasks outlined.

**Comment by kentonv** [comment]
@scottpurdy I think the tasks are:

* Fix portability issues that have regressed since 0.5.x. I suspect a bunch of stuff has broken, particularly related to the new test framework, but I won't know what until I try it. This is mostly "try to compile, find out what fails, fix it, repeat".
* Try to port the dynamic API. In other words, turn off lite mode and continue the above. I think you will need to make the compiler work since pycapnp invokes it dynamically (via the SchemaParser API), but you can avoid anything RPC-related.
* Add ifdefs or something to pycapnp to remove the RPC features when build on MSVC, since making RPC work on MSVC is going to be a huge undertaking that you probably don't want to get into.
* Figure out how to build pycapnp using MSVC against MSVC-built libcapnp.

**Comment by dariomt** [comment]
Hi!
Has there been any progress on this front?
I'd like to try the tasks listed above: any advice for a MSVC dev? Or can I just git clone and cmake -G "Visual Studio 14"? I don't see instructions for MSVC devs in the Installation page.

**Comment by kentonv** [comment]
Hi @dariomt,

You might find this thread interesting:

https://groups.google.com/d/msg/capnproto/DnlAgoiJGBI/7RnM8fqTAgAJ

**Comment by kentonv** [comment]
Version 0.6 -- to be released soon -- will fully support the dynamic API and RPC on MSVC 2017. Unfortunately MSVC still ICEs on the schema parser code, which blocks using this for Python...

I'm going to close this issue since it seems largely fixed but if someone wants to open an issue specifically for enabling pycapnp on MSVC, go ahead.

**Comment by Leandros** [comment]
Looks like that the LITE build is now broken. Declaring an object of `MallocMessageBuilder` throws a linker error: 

```
1> libcapnp.lib(layout.obj) : error LNK2019: unresolved external symbol "public: void __thiscall kj::ThrowOverflow::operator()(void)const " (??RThrowOverflow@kj@@QBEXXZ) referenced in function "public: class capnp::Data::Builder __thiscall capnp::_::PointerBuilder::getBlob<struct capnp::Data>(void const *,unsigned int)" (??$getBlob@UData@capnp@@@PointerBuilder@_@capnp@@QAE?AVBuilder@Data@2@PBXI@Z)
```

**Comment by kentonv** [comment]
@harrishancock ^

**Comment by kentonv** [comment]
@Leandros Can you file a new issue for this?

**Comment by harrishancock** [comment]
Will investigate.

**Comment by Leandros** [comment]
Duplicated as #576
